apps/drupal/src/Plugin/Preprocessor/PreprocessorManager.php

```diff
@@ -0,0 +1,509 @@
+   *   The entity object.
+   */
+  public static function getPageEntity() {
+    $page_entity = &drupal_static(__FUNCTION__, NULL);
+    if (isset($page_entity)) {
+      return $page_entity ?: NULL;
+    }
+
+    $route_match = \Drupal::routeMatch();
+    foreach ($route_match->getParameters() as $param) {
+      if ($param instanceof EntityInterface) {
+        $page_entity = $param;
+        break;
+      }
+    }
+
+    if (!isset($page_entity)) {
+      // Some routes don't properly define entity parameters.
+      // Thus, try to load them by its raw Id, if given.
+      $entity_type_manager = \Drupal::entityTypeManager();
+      $types = $entity_type_manager->getDefinitions();
+      foreach ($route_match->getParameters()->keys() as $param_key) {
+        if (!isset($types[$param_key])) {
+          continue;
+        }
+        if ($param = $route_match->getParameter($param_key)) {
+          if (is_string($param) || is_numeric($param)) {
+            try {
+              $page_entity = $entity_type_manager->getStorage($param_key)->load($param);
+            }
+            catch (\Exception $e) {
+            }
+          }
+          break;
+        }
+      }
+    }
+
+    if (!isset($page_entity) || !$page_entity->access('view')) {
+      $page_entity = FALSE;
+      return NULL;
+    }
+
+    return $page_entity;
+  }
+
+  /**
+   * Get all suggestions for the preprocessor.
+   *
+   * @param string $hook
+   *   The preprocess hook name. Example: 'node' for hook_preprocess_node.
+   * @param array $variables
+   *   The preprocess variables to alter.
+   *
+   * @return array
+   *   An array of preprocessor ids ordered with the most specific id first.
+   *   If $hook != $entity_type, then $entity_type is prefixed with $hook.
+   *   The following are from most specific to least specific when
+   *   For $hook === $entity_type: $prefix = $entity_type.
+   *   For $hook !== $entity_type: $prefix = {$hook}__{$entity_type}.
+   *   {$prefix}__group_context__{$group_bundle}__{$entity_bundle}__{$view_mode},
+   *   {$prefix}__{$entity_bundle}__{$view_mode},
+   *   {$prefix}__group_context__{$group_bundle}__{$entity_bundle},
+   *   {$prefix}__{$entity_bundle},
+   *   {$prefix}__group_context__{$group_bundle}__{$view_mode}
+   *   {$prefix}__{$view_mode},
+   *   {$prefix}__group_context__{$group_bundle}
+   *   {$prefix};
+   */
+  public static function getSuggestions($hook, array &$variables = []) {
+    // theme_get_suggestions().
+    // View mode.
+    $view_mode = !empty($variables['view_mode']) ? $variables['view_mode'] : NULL;
+
+    // Build suggestions based on the hook variable.
+    if (!empty($variables[$hook])) {
+      // Entity preprocessing.
+      $entity = NULL;
+      $entity_type = NULL;
+      $entity_bundle = NULL;
+      $entity_group_bundle = NULL;
+      $entity_view_mode = NULL;
+
+      if ($hook === 'page') {
+        // Page: Find the page entity.
+        $route_entity = static::getPageEntity();
```

Comment:
Probably out of scope for this PR but have you looked at if/how this could be used for preprocessing Views? Seems like this mostly relies on the page rendering a content entity and Views are config entities. Not a big deal but it would be cool if `template_preprocess_views_view()` could be a thing of the past.

Reply:
For sure. Mike did some much more complex uses of this and I'd love to see it expand and get tuned.